### ISSUE TITLE
any sample for webcam,thank you very much

### ISSUE THREAD

**Issue by xiang-zhe** [NVIDIA/VideoProcessingFramework#242]
Thanks for the great jobs.
In my case, i get video/frame from webcam using opencv, so i need copy it to gpu which costs a lot; after processing by model(in fact it costs little), i want the result to be numpy.narray dtype, so i'll copy it from gpu to cpu;
so fisrt, could i use vpf to  get video/frame from webcam without Host to Device copies;
second, is there a method copy data from gpu to cpu with high efficiency；

And i didn't find the describe details about how to use vpf, only some samples in github and brief introduction in `https://developer.nvidia.com/blog/vpf-hardware-accelerated-video-processing-framework-in-python/`
Any advice will be appreciated

**Comment by rarzumanyan** [comment]
Hi @xiang-zhe 

Most web cameras produce video in either mjpeg compressed format of yuv422 raw format.
E. g. this is how to check your webcam in Linux: (more information available [at this ffmpeg doc page](https://trac.ffmpeg.org/wiki/Capture/Webcam))
```bash
v4l2-ctl --list-devices
```

On my machine with Logitech web camera it outputs:
```
UVC Camera (usb-0000:00:14.0-13.3):
        /dev/video4
        /dev/video5
```

Then you check what video formats does it support:
```bash
ffmpeg -f v4l2 -list_formats all -i /dev/video4
```

Produces output:
```
[video4linux2,v4l2 @ 0x55dd5b6fc6c0] Raw       :     yuyv422 :           YUYV 4:2:2 : 640x480 160x120 176x144 320x176 320x240 352x288 432x240 544x288 640x360 752x416 800x448 800x600 864x480 960x544 960x720 1024x576 1184x656 1280x720 1280x960
[video4linux2,v4l2 @ 0x55dd5b6fc6c0] Compressed:       mjpeg :          Motion-JPEG : 640x480 160x120 176x144 320x176 320x240 352x288 432x240 544x288 640x360 752x416 800x448 800x600 864x480 960x544 960x720 1024x576 1184x656 1280x720 1280x960
```

The problem is mjpeg isn't supported by Nvdec: [Nvdec matrix support](https://developer.nvidia.com/video-encode-and-decode-gpu-support-matrix-new) and yuv422 files are too heavy to send over the USB in real time. So you either have to have a usb camera which outputs video in one of formats supported by Nvdec or use OpenCV as you do now.

>And i didn't find the describe details about how to use vpf, only some samples in github

Sample scripts illustrate most typical use cases and show how to use VPF. Unfortunately I don't have enough bandwidth to support the documentation especially when new features are added or some bugs are fixed.

>is there a method copy data from gpu to cpu with high efficiency

Yes, there's a `PySurfaceDownloader` class for that, it's usage is shown in `SampleDemuxDecode.py`.

**Comment by rarzumanyan** [comment]
Hi @xiang-zhe 

>so i NOT need to decode and just receive and convert to RGB array

Those format you've mentioned are raw image formats, they are not compressed. You don't need to decode them.

>how can i receive the raw video by vpf

If your camera supports raw yuv420p / nv12 / rgb output, you don't need VPF to obtain raw video frames from device. You can do that just fine with OpenCV because there's nothing a GPU can accelerate here - you just receive array of pixels over USB. 

As soon as you get your video frame as `uint8` numpy array with the help of OpenCV, you may upload it to GPU using `PyFrameUploader` and then export to `torch.tensor` for NN processing.

There are 2 ways to export VPF `Surface` to Pytorch tensor:

1. Using `PytorchNvCodec` module which utilizes Pytorch C++ API for that. It introduces the dependency against `torch` module in your python code.
2. By using `SurfacePlane.Export` method which is a simple CUDA DtoD memcopy. It doesn't introduce dependency against `torch` in your python code.

Both options are perfectly fine, just use whatever works best for you. One thing to take care of is the pixel format - some NN expect your tensor to be planar `float32` RGB, some prefer interleaved `float32` RGB, normalization range may be different and such. You can find some examples of `Surface` pre-processing in `SampleTorchResnet.py` and `SampleTensorRTResnet.py`.

>i didn't know how to convert a torch.tensor to a surface

Please take a look at `SamplePyTorch.py`:
https://github.com/NVIDIA/VideoProcessingFramework/blob/b896bef16a58e1183bcaa4406bd6b5024e890e50/SamplePyTorch.py#L73-L98

If you just want to download the content of your `torch.tensor` to numpy array, you don't need VPF for that, use `torch.tensor.cpu().numpy()` instead.

**Comment by xiang-zhe** [comment]
Hi @rarzumanyan ,
I was so careless that I didn't see it， 

> > i didn't know how to convert a torch.tensor to a surface
> 
> Please take a look at `SamplePyTorch.py`:
> 
> https://github.com/NVIDIA/VideoProcessingFramework/blob/b896bef16a58e1183bcaa4406bd6b5024e890e50/SamplePyTorch.py#L73-L98

in my case， ther’e‘re three  time-consuming processes，
1，numpy to gpu，
>0.016s, if using numpy.to("cuda"),  now using  PyFrameUploader (<0.004s)
2,  processing by NN,
<0.01s
3, tensor.cuda to numpy,
~ 0.02s, if using tensor.to("cpu"), so i hope to reduce it by VPF instead of using 

> torch.tensor.cpu().numpy()

i'll try it,thanks again!

**Comment by rarzumanyan** [comment]
Hi @xiang-zhe 
There's no need to guess, you can easily collect performance profile of your application.

When building VPF from source, opt-in `USE_NVTX` option to enable NVTX markers support. Then run your application under Nsight Systems profiler to see all the CUDA API calls and NVTX functions in application timelie.

Select Python interpreter, path to your script and it's arguments as target application.
Opt-in "Collect CUDA trace" and "Collect NVTX trace" options.
![image](https://user-images.githubusercontent.com/45193474/135237490-ea1760f6-44c1-42a3-a96a-a89663f2dd62.png)

You will see all VPF tasks in CUDA API calls in your app:
![image](https://user-images.githubusercontent.com/45193474/135237641-d5e7ee0f-6117-4e0d-abe0-9518332773f8.png)

P. S.
Please pull origin master first, I've added missing color conversion contexts in `SamplePyTorch.py`.

**Comment by xiang-zhe** [comment]
it's wired that now i use torch.to("cpu") and torch.to("cuda") is as fast as vpf, and the Occupancy rate of CPU is also low（15%）rather than 80+%。it makes me crazy
